src/draw/dma2d/lv_draw_dma2d.c

```diff
@@ -0,0 +1,569 @@
+
+    /* make sure the background area DMA2D is blending is up-to-date in main memory */
+    clean_cache(&cache_area);
+
+    uint32_t output_offset = (stride / cf_size) - w;
+    lv_draw_dma2d_configuration_t conf = {
+        .mode = LV_DRAW_DMA2D_MODE_MEMORY_TO_MEMORY_WITH_BLENDING_AND_FIXED_COLOR_FG,
+        .w = w,
+        .h = h,
+
+        .output_address = first_pixel,
+        .output_offset = output_offset,
+        .output_cf = output_cf,
+
+        .fg_color = lv_color_to_u32(color),
+        .fg_alpha_mode = 0x1, /* replace original fg alpha channel with fg_alpha */
+        .fg_alpha = opa,
+
+        .bg_address = first_pixel,
+        .bg_offset = output_offset,
+        .bg_cf = (lv_draw_dma2d_fgbg_cf_t) output_cf
+    };
+    configure_and_start_transfer(&conf, u);
+}
+
+static void configure_and_start_transfer(const lv_draw_dma2d_configuration_t * conf, lv_draw_dma2d_unit_t * u)
+{
+    /* number of lines register */
+    DMA2D->NLR = (conf->w << DMA2D_NLR_PL_Pos) | (conf->h << DMA2D_NLR_NL_Pos);
+
+    /* output */
+
+    /* output memory address register */
+    DMA2D->OMAR = (uint32_t)(uintptr_t) conf->output_address;
+    /* output offset register */
+    DMA2D->OOR = conf->output_offset;
+    /* output pixel format converter control register */
+    DMA2D->OPFCCR = ((uint32_t) conf->output_cf) << DMA2D_OPFCCR_CM_Pos;
+
+    /* Fill color. Only for mode LV_DRAW_DMA2D_MODE_REGISTER_TO_MEMORY */
+    DMA2D->OCOLR = conf->reg_to_mem_mode_color;
+
+    /* foreground */
+
+    /* foreground color. only for mem-to-mem with blending and fixed-color foreground */
+    DMA2D->FGCOLR = conf->fg_color;
+    /* foreground pixel format converter control register */
+    DMA2D->FGPFCCR = (conf->fg_alpha << DMA2D_FGPFCCR_ALPHA_Pos) | (conf->fg_alpha_mode << DMA2D_FGPFCCR_AM_Pos);
+
+    /* background */
+
+    DMA2D->BGMAR = (uint32_t)(uintptr_t) conf->bg_address;
+    DMA2D->BGOR = conf->bg_offset;
+    DMA2D->BGPFCCR = ((uint32_t) conf->bg_cf) << DMA2D_BGPFCCR_CM_Pos;
+
+
+    /* ensure the DMA2D register values are observed before the start transfer bit is set */
+    __DSB();
+
+    /* start the transfer (also set mode and enable transfer complete interrupt) */
+    DMA2D->CR = DMA2D_CR_START | (((uint32_t) conf->mode) << DMA2D_CR_MODE_Pos)
+#if LV_USE_DRAW_DMA2D_INTERRUPT
+                | DMA2D_CR_TCIE
+#endif
+                ;
+
+#if !LV_USE_DRAW_DMA2D_INTERRUPT
+#if LV_USE_OS
+    lv_result_t res = lv_thread_sync_signal(&u->thread_begin_polling_signal);
+    LV_ASSERT(res == LV_RESULT_OK);
+#else
+    await_transfer_completion();
+    post_transfer_tasks(u);
+#endif
+#endif
+}
+
+#if LV_USE_OS
+static void thread_cb(void * arg)
+{
+    lv_draw_dma2d_unit_t * u = arg;
+
+    while(1) {
+
+#if LV_USE_DRAW_DMA2D_INTERRUPT
+        lv_thread_sync_wait(&u->interrupt_signal);
+#else
+        lv_thread_sync_wait(&u->thread_begin_polling_signal);
+        await_transfer_completion();
```

Comment:
Is this scenario makes sense? It will spin a wile loop and we are the mercy of the OS to run other tasks while we are waiting here. Although it's a [low priority thread](https://github.com/lvgl/lvgl/pull/6691/files?diff=unified&w=0#diff-4a9425813cbac916944f0efcf3738f379cf07d3b44e1ed51558125096db3b61cR167) I wouldn't play gamble.

Reply:
If `lv_os.h` had something like `lv_os_sched_yield` or similar, we could check once and yield to other tasks. Scheduler yielding is available in most [RT]OSs, [even Linux](https://man7.org/linux/man-pages/man2/sched_yield.2.html) (although, unfortunately that isn't under the pthread umbrella :confused:).

I know the low priority isn't a reliable way to do it either. See my next comment in the main thread.